### PR TITLE
Kd_loss avg over tokens

### PR DESCRIPTION
**Pull request by moussaKam**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
Follow up to issue #1865
In ForwardKLWithChunkedOutputLoss  the losss is averaged over non-masked tokens instead of chunks
In this [notebook](https://colab.research.google.com/drive/1r6k1VGRKUK76aNTzSLlFrRfGd7BX3zk2?usp=sharing) I show an example.

I ran some experiments to compare the old and the new loss. I followed the tutorial [here](https://pytorch.org/torchtune/stable/tutorials/llama_kd_tutorial.html)

here's the two commands for the training
```
CUDA_VISIBLE_DEVICES=7 tune run knowledge_distillation_single_device --config llama3_2/knowledge_distillation_single_device kd_loss._component_=torchtune.modules.loss.OldForwardKLWithChunkedOutputLoss batch_size=16 gradient_accumulation_steps=8 enable_activation_checkpointing=False metric_logger.log_dir=/tmp/kd_output_old profiler.output_dir=/tmp/kd_output/profiling_outputs_old checkpointer.output_dir=/tmp/Llama-3.2-1B-Instruct_old/
```

```
CUDA_VISIBLE_DEVICES=6 tune run knowledge_distillation_single_device --config llama3_2/knowledge_distillation_single_device kd_loss._component_=torchtune.modules.loss.FixedForwardKLWithChunkedOutputLoss batch_size=16 gradient_accumulation_steps=8 enable_activation_checkpointing=False metric_logger.log_dir=/tmp/kd_output_fixed profiler.output_dir=/tmp/kd_output/profiling_outputs_fixed checkpointer.output_dir=/tmp/Llama-3.2-1B-Instruct_fixed/
```
The results were as follow: 
| Dataset        | Metric       | Old Loss| Fixed Loss|
|----------------|--------------|-----------|-----------|
| TruthfulQA     | mc1          | 0.2766    | 0.2778    |
| TruthfulQA     | mc2          | 0.4348    | 0.4381    |
| HellaSwag      | acc          | 0.4549    | 0.4542    |
| HellaSwag      | acc_norm     | 0.6103    | 0.6113    |
| CommonSenseQA  | acc          | 0.5553    | 0.5594    |

Please let me know if further work should be done

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by joecummings** [comment]
Thanks for the PR @moussaKam - I'll take a look later today and I've tagged @lindawangg, who implemented the first version of KD

File: torchtune/modules/loss/kd_losses.py [hunk 1]
```diff
@@ -78,7 +78,23 @@ def __init__(self, num_output_chunks: int = 8, ignore_index: int = -100):
         super().__init__()
         self.num_output_chunks = num_output_chunks
         self.ignore_index = ignore_index
-        self.fkl_loss = ForwardKLLoss(ignore_index)
+
+    def non_chunked_forward_kl_loss(
```
Comment (joecummings) [inline torchtune/modules/loss/kd_losses.py:82]: Why did you opt to re-create the `ForwardKLLoss` within the function instead of as another class that just returns the unnormalized calculation? I see a couple of options:

1. Re-use the `ForwardKLLoss` method but add a param called `normalize`, which will normalize the loss by non masked tokens
2. Create a new class called `ForwardKLLossUnnormalized` that returns the loss directly to be used in this implementation

All of these options assume we have the need to keep the original `ForwardKLLoss`. I don't know enough to say that we don't, but figured I'd ask if we might want to keep it or if we can just delete.

**Review by joecummings — COMMENTED** [review]
One design consideration, but looks good.

**Comment by moussaKam** [comment]
@ joecummings you're right it's better to use the existing class by adding a normalize param. Let me know if the modifications are good. 

As for `ForwardKLLoss` I think we should keep it because it's faster if enough memory is available.

File: torchtune/modules/loss/kd_losses.py [hunk 2]
```diff
@@ -50,6 +52,8 @@ def forward(
         prod_probs = torch.masked_fill(teacher_prob * student_logprob, inf_mask, 0)
         x = torch.sum(prod_probs, dim=-1).view(-1)
         mask = (labels != self.ignore_index).int()
+        if not normalize:
+            return -torch.sum(x * mask.view(-1), dim=0)
         if torch.sum(mask.view(-1), dim=0) == 0:
             return torch.tensor(0.0, device=x.device)
         return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
```
Comment (SalmanMohammadi) [inline torchtune/modules/loss/kd_losses.py:59]: FWIW we have `masked_mean` which we use in other parts of the codebase for this https://github.com/pytorch/torchtune/blob/17ba37d281548e5d60215f741888066717ad5b3e/torchtune/rlhf/rewards.py#L101 but not hugely fussed about whether it's used here.

**Review by lindawangg — APPROVED** [review]
Change makes sense to me. Thanks!

**Comment by xuwenxinedu** [comment]
I'm confused about the annotation 
in line 41 : labels (torch.Tensor): Ground truth labels of shape (batch_size, vocab_size) .
and line101: labels (torch.Tensor): Ground truth labels of shape (batch_size, num_tokens).
Why is it different?

**Comment by moussaKam** [comment]
Hi @xuwenxinedu looks like a typo, it should be num_tokens in both cases
